pgxn/neon/console_connector.c

```diff
@@ -0,0 +1,685 @@
+        pushJsonbValue(&state, WJB_BEGIN_ARRAY, NULL);
+
+        HASH_SEQ_STATUS status;
+        DbEntry *entry;
+        hash_seq_init(&status, RootTable.db_table);
+        while((entry = hash_seq_search(&status)) != NULL)
+        {
+            pushJsonbValue(&state, WJB_BEGIN_OBJECT, NULL);
+            PushKeyValue(&state, "op", entry->type == Op_Set ? "set" : "del");
+            PushKeyValue(&state, "name", entry->name);
+            if(entry->owner != InvalidOid)
+            {
+                PushKeyValue(&state, "owner", GetUserNameFromId(entry->owner, false));
+            }
+            if(entry->old_name[0] != '\0')
+            {
+                PushKeyValue(&state, "old_name", entry->old_name);
+            }
+            pushJsonbValue(&state, WJB_END_OBJECT, NULL);
+        }
+        pushJsonbValue(&state, WJB_END_ARRAY, NULL);
+    }
+
+    if(RootTable.role_table)
+    {
+        JsonbValue roles;
+        roles.type = jbvString;
+        roles.val.string.val = "roles";
+        roles.val.string.len = strlen(roles.val.string.val);
+        pushJsonbValue(&state, WJB_KEY, &roles);
+        pushJsonbValue(&state, WJB_BEGIN_ARRAY, NULL);
+
+        HASH_SEQ_STATUS status;
+        RoleEntry *entry;
+        hash_seq_init(&status, RootTable.role_table);
+        while((entry = hash_seq_search(&status)) != NULL)
+        {
+            pushJsonbValue(&state, WJB_BEGIN_OBJECT, NULL);
+            PushKeyValue(&state, "op", entry->type == Op_Set ? "set" : "del");
+            PushKeyValue(&state, "name", entry->name);
+            if(entry->password)
+            {
+                PushKeyValue(&state, "password", (char *)entry->password);
+            }
+            if(entry->old_name[0] != '\0')
+            {
+                PushKeyValue(&state, "old_name", entry->old_name);
+            }
+            pushJsonbValue(&state, WJB_END_OBJECT, NULL);
+        }
+        pushJsonbValue(&state, WJB_END_ARRAY, NULL);
+    }
+    JsonbValue *result = pushJsonbValue(&state, WJB_END_OBJECT, NULL);
+    Jsonb *jsonb = JsonbValueToJsonb(result);
+    return JsonbToCString(NULL, &jsonb->root, 0 /*estimated_len*/);
+}
+
+static void SendDeltasToConsole()
+{
+    if(!RootTable.db_table && !RootTable.role_table)
+        return;
+    if(!ConsoleURL)
+    {
+        elog(LOG, "ConsoleURL not set, skipping forwarding");
+        return;
+    }
+    if(!ForwardDDL)
+        return;
+
+    char *message = ConstructDeltaMessage();
+
+    curl_easy_setopt(CurlHandle, CURLOPT_CUSTOMREQUEST, "PATCH");
+    curl_easy_setopt(CurlHandle, CURLOPT_HTTPHEADER, ContentHeader);
+    curl_easy_setopt(CurlHandle, CURLOPT_POSTFIELDS, message);
+    curl_easy_setopt(CurlHandle, CURLOPT_URL, ConsoleURL);
+
+    const int num_retries = 5;
+    int curl_status;
+    for(int i = 0; i < num_retries; i++)
+    {
+        if((curl_status = curl_easy_perform(CurlHandle)) == 0)
+            break;
+        pg_usleep(1000 * 1000);
```

Comment:
Added logging. In regards to retrying if there's an HTTP error code, I'm not so sure. On the control plane side, it only returns an HTTP error code if the request was bad in some way, such as an empty password or wrong authentication. Since in these cases the request has no hope of succeeding anyway, I'm inclined to skip retrying if we get an error code.